src/rard/research/migrations/0061_order_in_book.py

```diff
@@ -0,0 +1,196 @@
+# Generated by Django 3.2 on 2023-01-23 11:20
+
+from django.db import migrations, models
+import itertools
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("research", "0060_add_reference_model"),
+    ]
+
+    def create_unknown_work(apps, schema_editor):
+        Antiquarian = apps.get_model("research", "Antiquarian")
+        Work = apps.get_model("research", "Work")
+
+        antiquarians = Antiquarian.objects.all()
+        for ant in antiquarians:
+            ant.works.add(Work.objects.create(name="Unknown Work", unknown=True))
+
+    def create_unknown_book(apps, schema_editor):
+        Work = apps.get_model("research", "Work")
+        works = Work.objects.all()
+        for work in works:
+            work.book_set.create(subtitle="Unknown Book", unknown=True)
+
+    def assign_links_to_unknown(apps, schema_editor):
+        worklinks = ["FragmentLink", "TestimoniumLink", "AppositumFragmentLink"]
+
+        for link_type in worklinks:
+            link_class = apps.get_model("research", link_type)
+            links = link_class.objects.all()
+            for link in links:
+                ant = link.antiquarian
+                if link.work is None:
+                    link.work = ant.works.get(unknown=True)
+                work = link.work
+                if link.book is None:
+                    link.book = work.book_set.get(unknown=True)
+                link.save()
+
+    def give_books_order(apps, schema_editor):
+        Work = apps.get_model("research", "Work")
+        for work in Work.objects.all():
+            books = work.book_set.all().order_by("unknown", "number")
+            for count, book in enumerate(books):
+                book.order = count
+                book.save()
+
+    def give_links_order_in_book(apps, schema_editor):
+        worklinks = ["FragmentLink", "TestimoniumLink", "AppositumFragmentLink"]
+
+        Work = apps.get_model("research", "Work")
+
+        def order_in_book_for_link_class(link_classname):
+            link_class = apps.get_model("research", link_classname)
+            for work in Work.objects.all():
+                related_links = link_class.objects.filter(work=work).order_by(
+                    "work_order"
+                )
+                for count, link in enumerate(related_links):
+                    link.order_in_book = count
+                    link.save()
```

Comment:
I think we need to be iterating through Books rather than Works here. As written, related_links will pick up all links belonging to the work, but for order_in_book, that's not what we want: For each Book we want to count from 1-N where N is the number of links belonging it.

**Note:** Still using work_order to order related_links though because we're starting off by deriving order in book from order in work.
 
```suggestion
        def order_in_book_for_link_class(link_classname):
            link_class = apps.get_model("research", link_classname)
            for book in Book.objects.all():
                related_links = link_class.objects.filter(book=book).order_by(
                    "work_order"
                )
                for count, link in enumerate(related_links):
                    link.order_in_book = count
                    link.save()
```